### PR TITLE
Update to Java 8

### PR DESCRIPTION
**Pull request by ham1**
Fixed previous botched PR. This relates to issue #96.

### CODE REVIEW

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java [hunk 1]
```diff
@@ -49,49 +50,39 @@ public Radio(WebElement wrappedElement) {
...
     public void selectByValue(String value) {
-        for (WebElement button : getButtons()) {
-            String buttonValue = button.getAttribute("value");
-            if (value.equals(buttonValue)) {
-                selectButton(button);
-                return;
-            }
-        }
+        Optional<WebElement> matchingButton = getButtons().stream()
+                .filter(b -> b.getAttribute("value").equals(value))
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java]: not NPE safe

should be `value.equals(buttonValue)` with precondition check, or `StringUtils.equals(value, btnVal)` if commons-lang3 available in classpath

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 2]
```diff
@@ -35,12 +27,16 @@ public Table(WebElement wrappedElement) {
...
+     *
+     * Multiple rows of heading elements, ({@code <tr>}), are flattened
+     * i.e. the second row, ({@code <tr>}), will follow the first, which can be
+     * misleading when the table uses {@code colspan} and {@code rowspan}.
      *
      * @return List with table heading elements.
      */
     public List<WebElement> getHeadings() {
-        return getWrappedElement().findElements(By.xpath(".//th"));
+        return getWrappedElement().findElements(By.cssSelector("th"));
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: should be in separate PR i think

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 3]
```diff
@@ -49,68 +45,70 @@ public Table(WebElement wrappedElement) {
...
      */
     public List<List<WebElement>> getRows() {
-        List<List<WebElement>> rows = new ArrayList<>();
-        List<WebElement> rowElements = getWrappedElement().findElements(By.xpath(".//tr"));
-        for (WebElement rowElement : rowElements) {
-            rows.add(rowElement.findElements(By.xpath(".//td")));
-        }
-        return rows;
+        return getWrappedElement()
+                .findElements(By.cssSelector("tr"))
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: changing selector type may produce unwanted side effects for some legacy browsers (which used a lot for some enterprise web apps). So please move change from xpath to css to separate PR

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 4]
```diff
@@ -49,68 +45,70 @@ public Table(WebElement wrappedElement) {
...
 
     /**
-     * Returns text values of table cell elements grouped by rows.
+     * Returns text values of table cell elements ({@code <td>}), grouped by rows.
      *
      * @return List where each item is text values of a table row.
      */
     public List<List<String>> getRowsAsString() {
-        return convert(getRows(), toListsConvertingEachItem(toTextValues()));
+        return getRows().stream()
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: think groupBy operator is existed for such operations

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 5]
```diff
@@ -49,68 +45,70 @@ public Table(WebElement wrappedElement) {
...
-                column.add(row.get(i));
-            }
-            columns.add(column);
-        }
-
-        return columns;
+        // assumes subsequent rows have same number of columns as
+        // this first row which contains <td>
+        List<WebElement> firstRow = getWrappedElement().findElements(
+                By.cssSelector("tr:nth-of-type(1) > td"));
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: working with prefetched columns should be faster (and more stable) than using findElements every time

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 6]
```diff
@@ -121,118 +119,38 @@ public WebElement getCellAt(int i, int j) {
...
-            for (String heading : headings) {
-                rowToHeadingsMap.put(heading, row.get(cellNumber));
-                cellNumber++;
-            }
-            rowsMappedToHeadings.add(rowToHeadingsMap);
-        }
-
-        return rowsMappedToHeadings;
+        return rows.stream()
+                .map(row -> row.stream().collect(toMap(e -> headings.get(row.indexOf(e)), e -> e)))
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: instead of `e -> e` there is a `Functions.identity()`

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/TextInput.java [hunk 7]
```diff
@@ -57,11 +59,8 @@ public String getText() {
             return getWrappedElement().getText();
         }
 
-        String enteredText = getWrappedElement().getAttribute("value");
-        if (enteredText == null) {
-            return "";
-        }
-        return enteredText;
+        Optional<String> enteredText = Optional.of(getWrappedElement().getAttribute("value"));
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/TextInput.java]: can be inlined i think

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java (hunk 5 above)
Comment (ham1) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: This is true, I will move this to a new method which you can chose which columns you would like, therefore should be faster if the number of columns you need is less than the total number.

**Comment by ham1** [comment]
Thanks for the comments @lanwen - I will have a look later and resubmit.

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java (hunk 1 above)
Comment (ham1) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java]: Good spot.

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java (hunk 4 above)
Comment (ham1) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: I could very well be wrong, but I thought ```groupingBy``` produced a ```Map<?, List<?>>``` - I don't see how it would help here?

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java [hunk 8]
```diff
@@ -49,49 +50,39 @@ public Radio(WebElement wrappedElement) {
...
-            }
-        }
+        Optional<WebElement> matchingButton = getButtons().stream()
+                .filter(b -> value.equals(b.getAttribute("value")))
+                .findFirst();
 
-        throw new NoSuchElementException(String.format("Cannot locate radio button with value: %s", value));
+        selectButton(matchingButton
+                .orElseThrow(() -> new NoSuchElementException(
+                        "Cannot locate radio button with value: " + value)));
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java]: `String.format("Cannot locate radio button with value: %s", value)` should be saved. Avoid string concatenation die to perfomance reasons

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java (hunk 4 above)
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: you can then getValues as list and it will be same result. But think its ok to leave it as you done

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java [hunk 9]
```diff
@@ -101,16 +104,30 @@ public Table(WebElement wrappedElement) {
...
 
     /**
-     * Returns text values of table cell elements grouped by columns.
+     * Returns table cell elements ({@code <td>}), grouped by columns.
+     *
+     * @return List where each item is a table column.
+     */
+    public List<WebElement> getColumnByIndex(int index) {
+        return getWrappedElement().findElements(
+                        By.cssSelector("tr > td:nth-of-type(" + index + ")"));
```
Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Table.java]: please use String.format to concatenate strings

**Comment by lanwen** [comment]
build fails due to old jdk preinstalled. Will fix that in a few days with qatools team

File: htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java (hunk 8 above)
Comment (ham1) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java]: The main reason I removed it was to improve readability as I think it's easier to read, especially since it only has one parameter which is at the end. Readability trumps any performance gain, especially when compared to the slowness of Selenium/browser interactions.

I can indeed keep the old String.format if it's for consistency or other reasons I've missed?

Comment (lanwen) [inline htmlelements-java/src/main/java/ru/yandex/qatools/htmlelements/element/Radio.java]: readability of  `+` is controversial point. Please save `format`

**Comment by lanwen** [comment]
retest this please

**Comment by lanwen** [comment]
@ham1 please squash commits, and i'll merge this

**Comment by emaks** [comment]
@lanwen :+1:

**Comment by ham1** [comment]
Merged latest changes and squashed commits into one - I can update the message if you have a preferred commit message structure or any suggested improvements.

**Comment by artkoshelev** [comment]
@ham1, thank you! i'll perform release at monday
